spec/integration_spec.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+require_relative "./spec_helper"
+require "fauxpaas"
+require "fauxpaas/cli/main"
+require "tmpdir"
+require "pathname"
+require "open3"
+
+module Fauxpaas
+
+  RSpec.describe "integration tests", integration: true do
+
+    describe "deploy" do
+
+      RSpec.shared_context "deploy setup" do |instance_name|
+        before(:all) do
+          @root = Pathname.new(Dir.tmpdir)/"fauxpaas"/"sandbox"/instance_name
+          `mkdir -p #{@root}`
+          CLI::Main.start(["deploy", instance_name])
+        end
+        after(:all) { `rm -rf #{@root}` }
+        let(:root) { @root }
+        let(:current_dir) { root/"current" }
+      end
+
+      # Expects
+      # let(:gem) { some gem name }
+      # let(:source) { some source file relative path }
+      RSpec.shared_examples "a successful deploy" do
+        it "the 'releases' dir exists" do
+          expect((root/"releases").exist?).to be true
+        end
+        it "the 'current' dir exists" do
+          expect((current_dir).exist?).to be true
+        end
+        it "installs unshared files" do
+          expect(File.read(current_dir/"some"/"dev"/"file.txt")).to eql("with some dev contents\n")
+        end
+        it "installs shared files" do
+          expect(File.read(current_dir/"some"/"shared"/"file.txt"))
+            .to eql("with some shared contents\n")
+        end
+        it "runs after_build commands" do
+          expect((current_dir/"eureka_2.txt").exist?).to be true
+          expect(File.read(current_dir/"eureka_1.txt")).to eql("eureka!\n")
+        end
+        it "installs the gems" do
+          # The expect {}-to-output syntax didn't like this test
+          Dir.chdir(current_dir) do
+            expect(`BUNDLE_GEMFILE=#{current_dir/"Gemfile"} bundle list`).to match(/#{gem}/)
+          end
+        end
+        it "installs the source files" do
+          expect((current_dir/source).exist?)
+            .to be true
+        end
+      end
+
+      context "without rails" do
+        include_context "deploy setup", "test-norails"
+        let(:gem) { "pry" }
+        let(:source) { Pathname.new("some_source_file.txt") }
+
+        include_examples "a successful deploy"
+      end
+
+      context "with rails" do
+        include_context "deploy setup", "test-rails"
+        let(:gem) { "rails" }
+        let(:source) { Pathname.new("config/environment.rb") }
+
+        include_examples "a successful deploy"
+
+        it "runs the migrations" do
+          expect(`sqlite3 #{current_dir/"db"/"production.sqlite3"} ".schema posts"`.chomp)
+            .to match(/CREATE TABLE "posts"/)
```

Comment:
This fails for me - could just be a different sqlite3 version or defaults:

```
       expected "CREATE TABLE IF NOT EXISTS \"posts\" (\"id\" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, \"name\" va...e\" varchar, \"content\" text, \"created_at\" datetime NOT NULL, \"updated_at\" datetime NOT NULL);" to match /CREATE TABLE "posts"/
       Diff:
       @@ -1,2 +1,2 @@
       -/CREATE TABLE "posts"/
       +"CREATE TABLE IF NOT EXISTS \"posts\" (\"id\" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, \"name\" varchar, \"title\" varchar, \"content\" text, \"created_at\" datetime NOT NULL, \"updated_at\" datetime NOT NULL)
```

I think changing it to `/CREATE TABLE (IF NOT EXISTS )?"posts"/` and likewise for `CREATE TABLE "things"` would be reasonable.